### ISSUE TITLE
Distorted audio reported only on Samsung S9/10 family of devices

### ISSUE THREAD

**Issue by derekleerock** [google/oboe#1204]
First of all, huge thanks to @dturner and @philburk (and your team/contributors!) for all of your work and active support on Oboe! I had the pleasure of meeting you both at Winter NAMM '18 after your presentation on Oboe and had been wanting to get into Android Audio dev for some time, and when that time came I knew that I wanted to use Oboe. I've seen some other Google repos not getting much support and am reassured by seeing how active you both are maintaining Oboe and I really appreciate that!

Android version(s): 10 (API 29), 11 (API 30)
Android device(s): Samsung S9, S9+, S10e, S10, S10+
Oboe version: 1.5.0 commit aead33c
App name used for testing: n/a
(Please try to reproduce the issue using the OboeTester or an Oboe sample.)
 
**Short description**
Users on these devices are reporting that the audio contains a painfully loud attack/percussive sound, distorted samples, changing the selected sample sound results in the same distorted audio, and that adjusting volume does not have any effect. The attack is so loud that they say it is painful to listen to using headphones. 

**Steps to reproduce**
I'm not sure I know how to reproduce this as it appears to only be happening on these specific devices (which I do not own), but here are the basics for how I am using Oboe, mostly influenced by RhythmGame sample code using the data and error callbacks, allowing Oboe to set the defaults for anything else:
```
    AudioStreamBuilder builder;
    builder.setFormat(oboe::AudioFormat::Float);
    builder.setPerformanceMode(oboe::PerformanceMode::LowLatency);
    builder.setSharingMode(oboe::SharingMode::Exclusive);
    builder.setSampleRateConversionQuality(oboe::SampleRateConversionQuality::Medium);
    builder.setChannelCount(oboe::ChannelCount::Mono);
    builder.setDataCallback(this);
    builder.setErrorCallback(this);
```

As this issue is only on recent versions of Android leveraging AAudio, I will omit the OpenSLES configuration that I have. 

WAV files are being read using the `WavStreamReader` and played back in the `onAudioReady()` method in time. 

I've spent a lot of time going through Oboe documentation and issues and also working with someone who has prior experience working with Oboe; I believe I've incorporated the latest code recommendations such as [using a shared pointer for the mAudioStream instead of a ManagedStream](https://github.com/google/oboe/issues/1126#issuecomment-739014632) and that my `onAudioReady()` method is not performing any blocking calls, among others. 

**Expected behavior**
Audio files are played as expected; audio responds appropriately when adjusting the volume (as appears to be the case on all other devices/models).

**Actual behavior**
Audio contains a loud attack, distorted samples, and adjusting volume does not have much effect, particularly at lower volumes and higher volumes. 

**Device**
Of the approximately 2,000 users who have downloaded and are using the app, I currently have 7 users reporting this behavior all using Samsung S9/S10 family of devices. With their help, I've been able to collect the audio stream properties from their devices in hoping that it would give some clues as per the following data:

<img width="1164" alt="Screen Shot 2021-02-28 at 4 33 48 PM" src="https://user-images.githubusercontent.com/3169260/109411095-e961c280-79e2-11eb-8f31-248a746e730f.png">

**Any additional context**
- Users have provided me with videos illustrating this issue. Please reference the following to see/hear what they are experiencing:
  - [Galaxy S9+, Android 10](https://drive.google.com/file/d/1enikDslwYyzm8Z9bwIh9DtfOhsP-n8AF/view?usp=sharing)
  - [Galaxy S10, Android 11](https://drive.google.com/file/d/1x2CXC6OTC-o0N5jS0BXgd16Djh2Y2jmJ/view?usp=sharing)
  - [Galaxy S10+, Android 11](https://drive.google.com/file/d/1le9tdDL-QAUN-cuobrK1tw3jxFgB3piA/view?usp=sharing)
- One user who I had asked for a video also [sent me a screen recording](https://drive.google.com/file/d/1chYsuIUMth203QRS6ENgAfpGU7QhdN47/view?usp=sharing) - and they told me that when they did the screen recording that the audio sounded correct. Therefore, this screen recording shows the proper audio as it is expected to sound. Maybe this is a hint? 
- I originally had some code which was adjusting the volume/gain of the sample ([according to recommendations that you had provided](https://github.com/google/oboe/issues/858#issuecomment-631594733)), and thought that this might be related, however even after removing these adjustments completely, the problem still persisted.
- I myself do not have a Samsung device that matches the specs that users reporting this issue have, therefore I cannot try to recreate this issue on my own unless I purchase one of these devices (and if possible I'd prefer not to do that). 
- I searched existing open and closed Oboe issues for anything that referenced "Samsung" and was surprised at how many results there were. I tried to go through all of these and ensure that my issue was not already covered somewhere else, but I may have overlooked something. 
- Admittedly, my C++ is rusty, so this could very likely be due to something that I am not doing correctly; however in my testing and in production this issue appears to only be limited to the specific devices that I've mentioned here. 

What other information can I collect to help diagnose this issue? I've tried to collect as much information as possible before opening an issue, but please let me know what else I can do to help.

**Comment by dturner** [comment]
Hi @derekleerock. Firstly, thanks for this fantastic bug report. This level of detail helps us a great deal to investigate the problem further. 

A few initial questions/observations: 

- Can you supply some examples of the WAV files you're using? What are the properties of these files in terms of format, channel count and sample rate?
- In your stream setup code you are specifying sample rate conversion, but you're not specifying a sample rate. This means that sample rate conversion will not be switched on, so when the stream's sample rate doesn't match the source data sample rate you'll get audio which is pitched up or down.
- According to @android-usb `WavStreamReader` doesn't support float but you're requesting floats for your stream format - could that be the problem? Btw you could use RhythmGame's `FFMpegExtractor` or `NDKExtractor` to extract your PCM data to a specified format/sample rate/channel count, although the former requires packaging FFmpeg. 

Anyway, let me know your thoughts on the above and we'll try to investigate further.

**Comment by derekleerock** [comment]
Thanks for your quick and kind response @dturner! I will dig into these points today and report back to you with additional info ASAP. 👍🏻

**Comment by derekleerock** [comment]
> Hi @derekleerock. Firstly, thanks for this fantastic bug report. This level of detail helps enormously when investigating and understanding the problem.

Thanks @dturner! There's a lot of potential data points, so it's hard to discern what the most relevant info would be to share. I really appreciate your support and suggestions on what other information is most useful to help discern this issue! 

> * Can you supply some examples of the WAV files you're using? What are the properties of these files in terms of format, channel count and sample rate?

Absolutely! I checked the output of a few utilities but thought the following output from running `soxi filename.wav` contained the most appropriate data. Please let me know if it would be better to use a different utility or if I'm missing any critical data points here.

For all WAV files that are being used, I've confirmed that they all have the following properties:

```
Channels       : 1
Sample Rate    : 48000
Precision      : 25-bit
Sample Encoding: 32-bit Floating Point PCM
```

Naturally, duration and file size are all different, but the above properties are all the same. [Here's one of the samples](https://drive.google.com/file/d/1MqLXteChiKtzS1CTLM2r1ANDTwW1I8Kx/view?usp=sharing) if you want to download and check it out.

> * In your stream setup code you are specifying sample rate conversion, but you're not specifying a sample rate. This means that sample rate conversion will not be switched on, so when the stream's sample rate doesn't match the source data sample rate you'll get audio which is pitched up or down.

Thanks for pointing this out! I do see one device in my sample data that has a different sample rate so this is likely impacting some users. I'll add this in - thank you!

> * According to @android-usb `WavStreamReader` doesn't support float but you're requesting floats for your stream format - could that be the problem? Btw you could use RhythmGame's `FFMpegExtractor` or `NDKExtractor` to extract your PCM data to a specified format/sample rate/channel count, although the former requires packaging FFmpeg.

This is a good question. I originally was using code inspired by RhythmGame's `FFMpegExtractor` and `NDKExtractor` and ran into issues with the WAV samples that I have which are single channel. At that point I changed to using the `WavStreamReader` to read in properties from the WAV file. Using this data, it's possible to create a new `AAssetDataSource` since the `WavStreamReader` allows access to the data that is needed, such as `getNumChannels()`, `getSampleRate()`, `getNumSampleFrames()`, `getSampleEncoding()`, `getBitsPerSample()`. It was then possible to fill the output buffer with the sample data from the WAV file (excluding the header) and set the output properties (channel count and sample rate) for the sample based on the reader. 

While it is possible to change the format from float to int, this feels like it would be unnecessary processing logic and additional complexity in the code since the samples are already in float format. Unless it's not possible to utilize the float data as it is, I think it would be simpler to continue to use float.

**Comment by philburk** [comment]
> While it is possible to change the format from float to int, this feels like it would be unnecessary
> processing logic and additional complexity in the code since the samples are already in float format. 

I agree.

> Unless it's not possible to utilize the float data as it is, I think it would be simpler to continue to use float.

Unfortunately it is not possible to use the float data as is. The WavStreamReader does not support Float. Note that the getFloatData() method only parses 16 bit short WAV data.

https://github.com/google/oboe/blob/108a2cd854cc298e348b73a2638abd42bc1a6c32/samples/parselib/src/main/cpp/wav/WavStreamReader.cpp#L128

I think what is happening is that when you read the sample data, nothing is happening. Then your app is playing uninitialized memory. It is not obvious because your sampled sound is a short tick. Uninitialized float data sounds like short noise burst. The data is volume scaled but it so far outside the range of -1 to +1 that it makes no difference.

The Samsung devices may behave differently in the way that they handle float clipping because the Samsung devices are using the new MMAP data path.

You have a few options:

1. Convert your samples to int16 using SOX, Audacity or other tool.
2. Add float support in WavStreamReader.cpp. You just have to move the data. No conversion required. If you do please share a pull request.
3. Wait for us to fix it.

**Comment by philburk** [comment]
I think we addressed the original issue so I am closing this. Let us know if you need more help on this.

**Comment by derekleerock** [comment]
Thanks for your kind support @philburk! No worries about closing this issue, but I would like to close the loop on this. 

> Unfortunately it is not possible to use the float data as is. The WavStreamReader does not support Float. Note that the getFloatData() method only parses 16 bit short WAV data.

Yes, I understand that `WavStreamReader` does not support float. My apologies for not being clearer in my previous response. 

We are only using `WavStreamReader` to get the number of samples, channel count, and bytes per sample. From there, we are then calculating the total PCM sample size in the file and reading it from the wav header offset (since WAV header size can very).

Here's the gist - credit goes to @ajaitx who assisted with this logic:

```
auto totalFileSize = AAsset_getLength(asset);
int8_t numberOfBytesPerSample = (reader.getBitsPerSample() / 8);
int32_t totalPCMSampleSize = reader.getNumSampleFrames() * outputProperties->channelCount * numberOfBytesPerSample;

int wavFileHeaderSize = (int) (totalFileSize - totalPCMSampleSize);
AAsset_seek(asset, SEEK_SET, wavFileHeaderSize);

auto *outputBuffer = new float_t[reader.getNumSampleFrames() * reader.getNumChannels()];
AAsset_read(asset, outputBuffer, totalPCMSampleSize);
```

From there, we can return an `AAssetDataSource` if the `reader.getSampleEncoding()` is `AudioEncoding::PCM_IEEEFLOAT` which contains the output properties for channel count and sample rate as retrieved from the `WavStreamReader`. 

> I think what is happening is that when you read the sample data, nothing is happening. Then your app is playing uninitialized memory. It is not obvious because your sampled sound is a short tick. Uninitialized float data sounds like short noise burst. The data is volume scaled but it so far outside the range of -1 to +1 that it makes no **difference.**

Are you suggesting that this is occurring on all devices, or just on some? 

The app is currently capable of playing 16 different WAV sample files, and between myself and about 2 dozen beta testers, some who have been testing on multiple devices, plus now users in production where the Play Console tells me that there are about 2,000 installs on active devices, have not experienced any audio issues at all. This includes confirming that proper audio is playing on Android versions as old as API 23 up to API 30, confirming both OpenSLES and AAudio paths.

Given this, the logic that we are using for reading in the WAV data as float doesn't feel like it is the issue here the behavior is only seen on certain device models. 

> The Samsung devices may behave differently in the way that they handle float clipping because the Samsung devices are using the new MMAP data path.

I wasn't aware of the MMAP data path previously, but [from a previous issue comment that you made](https://github.com/google/oboe/issues/1098#issuecomment-731588518), it appears as though the main phones which have this feature enabled does include Samsung and Pixel devices. We are also hearing of a different audio issue with some, but not all, Pixel devices. I didn't think it was related until you mentioned this. 

It appears as though [devices that are supporting MMAP only can support 16-bit integer data](https://source.android.com/devices/audio/aaudio#reporting-mmap-support):

```
<mixPort name="mmap_no_irq_out" role="source"
            flags="AUDIO_OUTPUT_FLAG_DIRECT|AUDIO_OUTPUT_FLAG_MMAP_NOIRQ">
            <profile name="" format="AUDIO_FORMAT_PCM_16_BIT"
                                samplingRates="48000"
                                channelMasks="AUDIO_CHANNEL_OUT_STEREO"/>
</mixPort>

<mixPort name="mmap_no_irq_in" role="sink" flags="AUDIO_INPUT_FLAG_MMAP_NOIRQ">
            <profile name="" format="AUDIO_FORMAT_PCM_16_BIT"
                                samplingRates="48000"
                                channelMasks="AUDIO_CHANNEL_IN_STEREO"/>
</mixPort>
```

If this is true, then my only option in order to support MMAP devices would be to process everything as int instead of float. 

> You have a few options:
> 1. Convert your samples to int16 using SOX, Audacity or other tool.

This sounds like the safest option - I've made this change and am waiting on test results from people who have impacted devices. 

> 2. Add float support in WavStreamReader.cpp. You just have to move the data. No conversion required. If you do please share a pull request.

The code that I referenced above is basically what we are using. I would be happy to make a PR but I don't have a lot of confidence in this area.. so long as someone is happy to help review it I can try to make that contribution. 

However, it's still not clear to me how this resolves the issue if the existing logic that we are using is not flawed. 

> 3. Wait for us to fix it.

Given that the app is in production and users are impacted, I'd like to take the quickest path to resolution for this and I don't know what your current priorities are and wouldn't want to have an impact on this. I am, however, curious what your thoughts are on the fix needed for this?

**Comment by derekleerock** [comment]
@philburk Quick update on this - my beta testers reported that the fix worked (I changed all samples and code to int16 instead of float) and I've released the app update to production and received confirmation from several impacted users that things are working well now! 🎉

Thanks so much to you and @dturner for your help! I don't know how I would've fixed this otherwise!

To close the loop on this - I can't quite recall the details of why I chose to go with float in the first place, but I believe it was an arbitrary decision as the samples that my sound engineer provided me with to use in the app just happened to be using float encoding. 

Per the comments that I left above, it seems that MMAP devices can only support Int16 data. If this is the case - then is it even possible to use float data with Oboe for a production app? I'm just wondering how I should approach future use cases, if other people can learn from this experience so they don't make this same mistake, etc? Any additional insights you can share here would be great. 

I'd also be happy to make a PR to help improve documentation if you feel that would be helpful, too.

**Comment by philburk** [comment]
> my beta testers reported that the fix worked (I changed all samples and code to int16 instead of float) and
> I've released the app update to production and received confirmation from several impacted users that
> things are working well now!

That is great news but I cannot explain it. AAudio converts everything to float internally and then converts it to Int16 before write the data to the DSP buffer for MMAP. So there should not be any device dependencies based on what the application is using.

Oboe does support float.
